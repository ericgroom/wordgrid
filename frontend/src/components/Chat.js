import React from "react";
import ReactDOM from "react-dom";
import { connect } from "react-redux";
import ChatWindow from "./ChatWindow";

class Chat extends React.Component {
  state = {
    message: "",
    show: false
  };
  chatContainer = (function() {
    const el = document.createElement("div");
    el.id = "chat";
    return el;
  })();
  componentDidMount() {
    let portalDiv = document.getElementById("portal");
    if (!portalDiv) {
      portalDiv = document.createElement("div");
      portalDiv.id = "portal";
      document.getElementsByTagName("body")[0].appendChild(portalDiv);
    }
    portalDiv.appendChild(this.chatContainer);
  }
  componentWillUnmount() {
    const portalDiv = document.getElementById("portal");
    portalDiv.removeChild(this.chatContainer);
  }
  sendMessage = message => {
    this.props.sendMessage(message);
  };
  handleChange = e => {
    this.setState({ message: e.target.value });
  };
  render() {
    const { messages } = this.props;
    return ReactDOM.createPortal(
      <ChatWindow sendMessage={this.sendMessage} messages={messages} />,
      this.chatContainer
    );
  }
}

const mapStateToProps = state => ({
  messages: state.messages.messages
});

const mapDispatchToProps = dispatch => ({
  sendMessage: message => dispatch({ type: "SEND_MESSAGE", message })
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Chat);
