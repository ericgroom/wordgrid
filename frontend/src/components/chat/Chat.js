import React from "react";
import ReactDOM from "react-dom";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import ChatWindow from "./ChatWindow";

/**
 * Container component for chat. Interfaces with redux as well as mounts
 * ChatWindow as a portal
 */
export class Chat extends React.Component {
  static propTypes = {
    /** List of messages to display */
    messages: PropTypes.arrayOf(
      PropTypes.shape({
        sender: PropTypes.string,
        message: PropTypes.string
      })
    ),
    /** Function to call when the user sends a message */
    sendMessage: PropTypes.func.isRequired
  };
  chatContainer = (function() {
    const el = document.createElement("div");
    el.id = "chat";
    return el;
  })();
  componentDidMount() {
    let portalDiv = document.getElementById("portal");
    if (!portalDiv) {
      portalDiv = document.createElement("div");
      portalDiv.id = "portal";
      document.getElementsByTagName("body")[0].appendChild(portalDiv);
    }
    portalDiv.appendChild(this.chatContainer);
  }
  componentWillUnmount() {
    const portalDiv = document.getElementById("portal");
    portalDiv.removeChild(this.chatContainer);
  }
  /**
   * Calls props.sendMessage with message content.
   *
   * @param {string} message message text
   */
  sendMessage = message => {
    if (message && message.length > 0) {
      this.props.sendMessage(message);
    }
  };
  handleChange = e => {
    this.setState({ message: e.target.value });
  };
  render() {
    const { messages } = this.props;
    return ReactDOM.createPortal(
      <ChatWindow sendMessage={this.sendMessage} messages={messages} />,
      this.chatContainer
    );
  }
}

const mapStateToProps = state => ({
  messages: state.messages.messages
});

const mapDispatchToProps = dispatch => ({
  sendMessage: message => dispatch({ type: "SEND_MESSAGE", message })
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Chat);
