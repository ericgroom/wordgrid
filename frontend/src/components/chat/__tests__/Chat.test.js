import React from "react";
import { shallow } from "enzyme";
import { Chat } from "../Chat";

describe("<Chat />", () => {
  it("renders without crashing", () => {
    const wrapper = shallow(<Chat sendMessage={() => {}} />);
    expect(wrapper).toMatchSnapshot();
  });
});
