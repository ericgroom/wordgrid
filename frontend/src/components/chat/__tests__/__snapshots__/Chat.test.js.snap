// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Chat /> renders without crashing 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Chat
    sendMessage={[Function]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "portal",
    "props": Object {
      "children": <ChatWindow
        sendMessage={[Function]}
      />,
      "containerInfo": <div
        id="chat"
      />,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "messages": undefined,
        "sendMessage": [Function],
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
    "type": Symbol(react.portal),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "portal",
      "props": Object {
        "children": <ChatWindow
          sendMessage={[Function]}
        />,
        "containerInfo": <div
          id="chat"
        />,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "messages": undefined,
          "sendMessage": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      "type": Symbol(react.portal),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
